Demo drivers

ranechita opened the pull request.

ranechita commented:
v2

- ran astyle.sh

mchindri commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -50,22 +50,34 @@
 struct adc_demo_init_param {
+	/* NUmber of channels*/
+	uint32_t channel_no;
+	/**buffer for communication between adc&dac, according to appropriate channel*/
```
I think this can be described as array of buffers for each channel.

ranechita replied:
Did you mean in the comment description?

mchindri replied:
yes

mchindri commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -50,22 +50,34 @@
+	/* NUmber of channels*/
+	uint32_t channel_no;
+	/**buffer for communication between adc&dac, according to appropriate channel*/
+	uint16_t **loopback;
```
loopback_buffers maybe :)

mchindri commented on projects/iio_demo/src/app/main.c:
```diff
@@ -194,12 +213,48 @@ int main(void)
+		.size = MAX_SIZE_BASE_ADDR
+	};
+
+	uint16_t** buf = (uint16_t**)malloc(10*sizeof(uint16_t*));
```
in iio_adc_demo.h : #define ADC_NUM_CHANNELS or something
and here: uint16_t loopback_buff[ADC_NUM_CHANNELS][MAX_SAMPLES_PER_CHANNEL]

mchindri commented on projects/iio_demo/src/app/main.c:
```diff
@@ -194,12 +213,48 @@ int main(void)
 		.size = MAX_SIZE_BASE_ADDR
 	};
 
+	struct iio_data_buffer buff = {
```
adc_buff and dac_buff is a better naming

mchindri commented on drivers/adc/adc_demo/iio_adc_demo.h:
```diff
@@ -42,10 +42,29 @@
+	.attributes = NULL,	\
+	.debug_attributes = NULL,	\
+	.buffer_attributes = NULL,	\
+	.prepare_transfer = init_adc_channels,	\
```
update_active_adc_channels should go here

ranechita replied:
the number of channels is now configurable from the init param, and that function retrieves the actual number of active channels

mchindri replied:
prepare_transfer shuld only update the mask with the activ channels, it is how iio works.

When selecting from the oscilloscope channel 0 the mask will be 0x00000001,  when selecting channel 1, it will be 0x00000002 and if both will be 0x00000003

ranechita replied:
My bad, I found this out and I also had to modify some other things and now it has update_adc_channels

ranechita commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +108,79 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
 	return SUCCESS;
 }
 
+int32_t adc_get_number_of_channels(void* dev)
```
I forgot to delete it, it was from a previous approach. Removed it now

mchindri replied:
for all

ranechita commented on drivers/dac/dac_demo/dac_demo.h:
```diff
@@ -50,22 +50,32 @@
-#define MAX_REG_ADDR		16
+#define MAX_DAC_ADDR		16
+#define DEFAULT_LOCAL_SAMPLES 128
+#define SINE_VECTOR_SIZE 128
```
removed it

ranechita commented:
v3

- init param field loopback now named loopback_buffers
- renamed iio_Data_buffers buff&buff1 to adc_buff and dac_buff
- the buffer from main `uint16_t** buf = (uint16_t**)malloc(10*sizeof(uint16_t*));` -> `uint16_t loopback_buff[MAX_ADC_CHANNELS][MAX_SAMPLES_PER_CHANNEL]`
- used array size instead of number `buff[k] = sine_lut[k%128];` -> `buff[k] = sine_lut[k % ARRAY_SIZE(sine_lut)]`
- removed unused function `int32_t adc_get_number_of_channels(void* dev)`
- removed unused define for sine_lut size

ranechita commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -77,6 +89,14 @@ int32_t adc_demo_init(struct adc_demo_desc **desc,
 
 int32_t adc_demo_remove(struct adc_demo_desc *desc);
 
+int32_t adc_get_number_of_channels(void* dev);
```
was this a general remark (for functions like close_dac_channels,etc) or was it specific for this function?

adrimbarean commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -50,22 +50,34 @@
+	uint32_t active_ch;
+	/** buffer for communication between adc&dac, according to channel*/
+	uint16_t **loopback;
+//	/* NUmber of channels*/
+//	uint32_t channel_no;
```
Is commented code needed?

ranechita replied:
i removed it

ranechita commented:
v4

- removed unnecessary comment in `adc_demo_desc` in adc_demo.h
- shortened commit message

ranechita commented:
v5

- moved `if(!dev) return -ENODEV;` after variable definition

mirceacaprioru commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +112,144 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+	}
+	*new_idx = last_idx;
+
+	return 1;
```
should be consistent in returns (return 0 return 1  or return SUCCESS return -ENODEV) 

If there are no more channels you can just return -ENOENT or something like that. There are just a few functions that return something positive like snprintf which returns the length of the string. In this case i would go with typical return SUCCESS or return and error if there is no next ch index but this is not a strong request so i'll leave it up to you.

mchindri replied:
Well, this may be for me, I first wrote this function in iio_demo: 
The reason to return 1 was just to be more human readable in:
while(get_next_ch_idx ...
instead of while (0 == get_next_ch_idx)

mchindri replied:
any way, I think we can add this to util.h beacause the same logic may be used in new iio drivers

mirceacaprioru commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +112,144 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+ * @param len - Maximum length of value to be stored in buf.
+ * @param channel - Channel properties.
+ * @param attr_id - Attribute ID
+ * @return Length of chars written in buf, or negative value on failure.
```
this is a good example of a positive return related to chars written in a buffer.

ranechita commented:
v6

- re-added [k % ARRAY_SIZE(sine_lut)];
- modified `.prepare_transfer =` to update_adc_channels and update_dac_channels
- modified the init_adc_channels and init_dac_channels to work with osc
- added calls in main for init_adc_channels and init_dac_channels, to initialize the channels before adding the devices
-  moified from `void* dev ` to `struct adc_demo_desc* desc` in init_adc_channels and init_dac_channels

ranechita commented:
v7

- fixed issues with attributes, in get/set attributes functions and in init_adc/dac_channels functions
- ran astyle.sh

mchindri commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -45,27 +45,46 @@
-#define MAX_REG_ADDR		16
+#define MAX_ADC_ADDR		16
+#define DEFAULT_LOCAL_SAMPLES 128
+#define SINE_VECTOR_SIZE 128
```
Not used

mchindri commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -45,27 +45,46 @@
 /******************************************************************************/
 
 #include <stdint.h>
+#include "iio_types.h"
```
Not needed here

mchindri commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -45,27 +45,46 @@
+	/** Active channel**/
+	uint32_t active_ch;
+	/** Array of buffers for each channel*/
+	uint16_t **loopback;
```
Maybe you can rename this to to be consistent with the one in init param

mchindri commented on drivers/dac/dac_demo/dac_demo.c:
```diff
@@ -86,12 +91,164 @@ int32_t dac_demo_remove(struct dac_demo_desc *desc)
  * @param readval- the value read from register
  * @return SUCCESS in case of success, negative error code otherwise.
-*******************************************************************************/
+**************************************************************************/
```
These changes are not needed

mchindri commented on drivers/adc/adc_demo/iio_adc_demo.h:
```diff
@@ -42,10 +42,45 @@
+}
+
+
+int32_t init_adc_channels(struct adc_demo_desc* dev, uint32_t mask);
```
I would initialize as here https://github.com/analogdevicesinc/no-OS/blob/60c83ad3b11fe54de086746dce6607301399316b/iio/iio_adxrs290/iio_adxrs290.h#L144 for examples.
iio_adc_demo and iio_dac_demo should serve as example for new drivers and be as simple as possible. 
Therefore I think is better to use a fixed number of channels because the chips we are creating drivers for have a fixed number too.

dbogdan replied:
Having the option to choose a different number of channels would help us testing the entire infrastructure much easier - we want to make sure everything works even if for example a device has 1 channel or has 16 channels.

ranechita commented:
v8

- fixed styling issues
- removed unnecessary includes and defines

ranechita commented:
v9

- added a channel list of 16 channels
- device number of channels configurable by modifying ADC_CHANNEL_NO/DAC_CHANNEL_NO

mchindri commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -41,11 +41,34 @@
+#include <string.h>
+#include <inttypes.h>
+#include "iio_types.h"
+#include "iio_adc_demo.h"
```
iio_adc_demo.h shouldn't be needed here (the driver doesn't need to know about iio)

mchindri commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +112,163 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+
+	if(desc->loopback_buffers == NULL) {
+		//default sin function
+		for(k = 0 ; k < samples; k++)
+			buff[k] = sine_lut[k % ARRAY_SIZE(sine_lut)];
```
```suggestion
                int offset_per_ch = ARRAY_SIZE(sine_lut) / TOTAL_CHANNELS;
		for(int i = 0; i < samples; i++) {
		     while(get_next_ch_idx(desc->active_ch,ch,&ch))
			buff[k++] = sine_lut[(i + ch * offset_per_ch ) % ARRAY_SIZE(sine_lut)];
```
This will print a sin with an offset for each channel

mchindri commented on drivers/adc/adc_demo/adc_demo.h:
```diff
@@ -50,22 +50,39 @@
+	/** Demo global device attribute */
+	uint32_t adc_global_attr;
+	/** Demo device channel attribute */
+	uint32_t adc_ch_attr;
```
uint32_t adc_ch_attr[TOTAL_CHANNELS]

This way you cand change an attribute for each channel.
And at line 263: desc->adc_ch_attr[channel->ch_num] = value;

mchindri commented on drivers/adc/adc_demo/iio_adc_demo.c:
```diff
@@ -0,0 +1,97 @@
+#include "iio_types.h"
+#include "error.h"
+
+/*struct iio_channel iio_adc_channels[MAX_NR_CHANNELS];
```
You commented this by mistake I assume

ranechita replied:
This file was supposed to be deleted, but i deleted it only form sdk and it remained in the drivers folder

ranechita commented:
v10

- deleted iio_adc_demo.c and iio_dac_demo.c
- added offset for each channel for sin
- moved `enum iio_adc_demo_attributes { ADC_CHANNEL_ATTR, ADC_GLOBAL_ATTR, };` to adc_demo.h and replaced the include iio_adc_demo.h with adc_demo.h

ranechita commented:
v11

- added different value for each channel attribute

amiclaus commented on drivers/adc/adc_demo/iio_adc_demo.h:
```diff
@@ -42,10 +42,86 @@
+	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write	\
+}
+
+//static struct iio_device adc_demo_iio_descriptor = ADC_DEMO_DEV(MODIFIED_CHANNEL_VALUE);
```
are these 2 commented lines on purpose in this file?

ranechita replied:
It is commented on purpose, i left them because i was unsure if the initialization of the number of channels is going to be done through MODIFIED_CHANNEL_VALUE in main or if I should try a different approach

amiclaus commented on drivers/dac/dac_demo/iio_dac_demo.h:
```diff
@@ -42,10 +42,82 @@
+	IIO_DEMO_DAC_CHANNEL(15)
+};
+
+//static struct iio_device dac_demo_iio_descriptor = DAC_DEMO_DEV(2);
```
same mention as above. Is this line needed, since it is commented?

amiclaus commented on projects/iio_demo/src.mk:
```diff
@@ -8,13 +8,26 @@ SRCS +=	$(NO-OS)/util/xml.c						\
+		$(DRIVERS)/dac/dac_demo/iio_dac_demo.c				\
+		$(DRIVERS)/adc/adc_demo/iio_adc_demo.c				\
+
+
```
nitpick: extra line

amiclaus commented on projects/iio_demo/src.mk:
```diff
@@ -8,13 +8,26 @@ SRCS +=	$(NO-OS)/util/xml.c						\
 	$(INCLUDE)/util.h						\
 	$(INCLUDE)/error.h
 
+
```
nitpick: extra line

amiclaus commented on projects/iio_demo/src.mk:
```diff
@@ -72,4 +85,3 @@ INCS += $(PROJECT)/src/hal/stm32/main.h \
 
 LSCRIPT = $(PROJECT)/src/linker/STM32F446RETX_FLASH.ld
 endif
-
```
nitpick: I think this modification is unrelated to the PR.

amiclaus commented on projects/iio_demo/src/app/main.c:
```diff
@@ -146,6 +149,9 @@ int32_t platform_init()
 	return 0;
 }
 
+//uint16_t buf[2][1000];
```
this can be removed.

amiclaus commented on projects/iio_demo/src/app/main.c:
```diff
@@ -194,12 +212,48 @@ int main(void)
+	status = dac_demo_init(&dac_desc, &dac_init_par);
+	if (status != SUCCESS)
+		return status;
+	//MODIFIED_CHANNEL_VALUE = 16;
```
is this still required?

amiclaus commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +113,166 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+		//default sin function
+		int offset_per_ch = ARRAY_SIZE(sine_lut) / TOTAL_CHANNEL_NO;
+		for(int i = 0; i < samples; i++) {
+			while(get_next_ch_idx(desc->active_ch,ch,&ch))
```
nitpick:
`get_next_ch_idx(desc->active_ch,ch,&ch)` -> `get_next_ch_idx(desc->active_ch, ch, &ch)`
check throughout the entire code.

mchindri commented on drivers/adc/adc_demo/iio_adc_demo.h:
```diff
@@ -42,10 +42,86 @@
-	.debug_reg_read = (int32_t (*)()) adc_demo_reg_read,
-	.debug_reg_write = (int32_t (*)()) adc_demo_reg_write
+#define DEFAULT_CHANNEL_NO 2
+#define ADC_CHANNEL_NO changeable_channel_number
```
I would move this to adc_demo.h as:
#infdef MODIFIED_CHANNEL_VALUE
#define ADC_CHANNEL_NO 2
#else
#define ADC_CHANNEL_NO 16
#endif

ranechita replied:
I wrote them this way so that it can be used with values other than 2 and 16

amiclaus replied:
then I think you can do something like this:

**#ifndef ADC_CHANNEL_NO**
#infdef MODIFIED_CHANNEL_VALUE
#define ADC_CHANNEL_NO 2
#else
#define ADC_CHANNEL_NO 16
#endif
**#endif**

This allows you to define custom `ADC_CHANNEL_NO` macro inside your main app directly.

And you can avoid passing static variables to macros, which I don't think is a good practice.

mchindri commented on projects/iio_demo/src/app/main.c:
```diff
@@ -146,6 +149,9 @@ int32_t platform_init()
 }
 
+//uint16_t buf[2][1000];
+uint16_t buf[DEFAULT_CHANNEL_NO][MAX_SAMPLES_PER_CHANNEL];
```
use ADC_CHANNEL_NO

mchindri commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +113,166 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+
+	for(int i = 0; i < samples; i++) {
+		while(get_next_ch_idx(desc->active_ch,ch,&ch))
+			buff[k++] = desc->loopback_buffers[ch][i % DEFAULT_LOCAL_SAMPLES];
```
I would define DEFAULT_LOCAL_SAMPLES as a field in init param as: ch_buffer_size

mchindri commented on drivers/dac/dac_demo/dac_demo.c:
```diff
@@ -37,25 +37,28 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *******************************************************************************/
 
-/******************************************************************************/
```
Can you remove comments modifications?

mchindri commented on drivers/dac/dac_demo/dac_demo.c:
```diff
@@ -86,12 +93,164 @@ int32_t dac_demo_remove(struct dac_demo_desc *desc)
+		while (get_next_ch_idx(desc->active_ch, ch, &ch))
+			desc->loopback_buffers[ch][i % DEFAULT_LOCAL_SAMPLES] = buff[k++];
+
+
```
extra line

mchindri commented on projects/iio_demo/src/app/main.c:
```diff
@@ -194,12 +212,48 @@ int main(void)
+	struct iio_device dac_demo_iio_descriptor = DAC_DEMO_DEV(dac_channels);
+
 	struct iio_app_device devices[] = {
-		IIO_APP_DEVICE("demo_device_output", iio_demo_out_desc,
```
When removing this you also have to remove remaining old iio_demo structures.
Anyway, maybe you can just keep old ones in this commit and in an other commit remove this an all related with old iio_demo

ranechita commented:
v12

- removed extra lines
- corrected `get_next_ch_idx(desc->active_ch,ch,&ch)` to `get_next_ch_idx(desc->active_ch, ch, &ch)`
- replaced DEFAULT_CHANNEL_NO with ADC_CHANNEL_NO
- removed some unnecessary comments
